emle/models.py

```diff
@@ -1210,3 +1223,394 @@ def forward(self, atomic_numbers, charges_mm, xyz_qm, xyz_mm):
+
+        if atomic_numbers is not None:
+            if not isinstance(atomic_numbers, _torch.Tensor):
+                raise TypeError("'atomic_numbers' must be of type 'torch.Tensor'")
+            # Check that they are integers.
+            if atomic_numbers.dtype is not _torch.int64:
+                raise ValueError("'atomic_numbers' must be of dtype 'torch.int64'")
+            self._atomic_numbers = atomic_numbers.to(device)
+        else:
+            self._atomic_numbers = None
+
+        # Call the base class constructor.
+        super().__init__(
+            alpha_mode=alpha_mode,
+            device=device,
+            dtype=dtype,
+            create_aev_calculator=True,
+        )
+
+        # Load the MACE model.
+        if mace_model is not None:
+            if mace_model.startswith("mace-off23"):
+                size = mace_model.split("-")[-1]
+                assert (
+                    size in ["small", "medium", "large"]
+                ), f"Unsupported MACE model: '{mace_model}'. Available MACE-OFF23 models are 'mace-off23-small', 'mace-off23-medium', 'mace-off23-large'"
+                self._mace = _mace_off(model=size, device=device, return_raw_model=True)
+            else:
+                # Assuming that the model is a local model.
+                if _os.path.exists(mace_model):
+                    self._mace = _torch.load(mace_model, map_location="device")
+                else:
+                    raise FileNotFoundError(f"MACE model file not found: {mace_model}")
+        else:
+            # If no MACE model is provided, use the default MACE-OFF23(S) model.
+            self._mace = _mace_off(model="small", device=device, return_raw_model=True)
+
+        # Compile the model.
+        self._mace = _e3nn_jit.compile(self._mace).to(self._dtype)
+
+        # Create the z_table of the MACE model.
+        self._z_table = [int(z.item()) for z in self._mace.atomic_numbers]
+
+        if self._atomic_numbers is not None:
+            # Get the node attributes.
+            node_attrs = self._get_node_attrs(self._atomic_numbers)
+            self.register_buffer("ptr", _torch.tensor([0, node_attrs.shape[0]], dtype=_torch.long, requires_grad=False),)
+            self.register_buffer("node_attrs", node_attrs.to(self._dtype))
+            self.register_buffer("batch",_torch.zeros(node_attrs.shape[0], dtype=_torch.long, requires_grad=False),)
+
+            print("Node attributes:", node_attrs.shape)
+            print("Batch:", self.batch.shape)
+            print("Ptr:", self.ptr.shape)
+            print(node_attrs)
+
+        # No PBCs for now.
+        self.register_buffer("pbc",_torch.tensor([False, False, False], dtype=_torch.bool, requires_grad=False),)
+        self.register_buffer("cell", _torch.zeros((3, 3), dtype=self._dtype, requires_grad=False))
+
+        # Assign a tensor attribute that can be used for assigning the AEVs.
+        self._aev_computer._aev = _torch.empty(0, device=device)
```

Comment:
Is this needed? It should only be needed for `ANI2xEMLE` when hooking the internal AEVComputer of torchani.

Reply:
Removed. I assumed the tensor had to be initialized before any forward pass.

Reply:
When not using the forward hook we just call the AEVComputer directly. The issue with the hook is that I needed some way of capturing the AEVs, so store it as a attribute of the AEVComputer inside the hook. TorchScript only lets you do this if the attribute exists before the hook is called, hence needed to initialise it.